### PR TITLE
Implement precomputing of datashader input data

### PR DESCRIPTION
**Pull request by philippjfr**
An initial go at https://github.com/ioam/holoviews/issues/2196. It allows datashader operation to precompute results, which can be used by the operation. For now only implemented for TriMesh rasterization but should be straightforward enough to do for other aggregation (although unnecessary for rasters).

I'm still considering how to improve this further. I think memoizing the precomputed results based on the element ID would be sufficient but will continue thinking about it.

### CODE REVIEW

**Comment by jlstevens** [comment]
I know it isn't ready for review yet but I will say that at a glance, this looks like a very reasonable approach.

**Comment by philippjfr** [comment]
I'm now using the ``element._plot_id`` to memoize the precomputed data on as that's specifically used in the plotting code to denote static data.

File: holoviews/operation/datashader.py [hunk 1]
```diff
@@ -502,27 +519,41 @@ class trimesh_rasterize(aggregate):
                                          objects=['bilinear', None], doc="""
         The interpolation method to apply during rasterization.""")
 
+    def _precompute(self, element):
+        from datashader.utils import mesh
```
Comment (jlstevens) [inline holoviews/operation/datashader.py]: Why the inline import?

File: holoviews/operation/datashader.py [hunk 2]
```diff
@@ -567,21 +598,27 @@ def _process(self, element, key=None):
...
                                and not issubclass(x.type, Image))
         element = element.map(dsrasterize, predicate)
+        self._precomputed = trirasterize._precomputed
 
         # Rasterize other Dataset types
         predicate = lambda x: (isinstance(x, Dataset) and
                                (not isinstance(x, Image) or x in imgs))
         element = element.map(dsrasterize, predicate)
+        self._precomputed = trirasterize._precomputed
+
```
Comment (jlstevens) [inline holoviews/operation/datashader.py]: If any more of these appear I would recommend a context manager.

File: holoviews/operation/datashader.py [hunk 3]
```diff
@@ -89,6 +90,17 @@ class ResamplingOperation(Operation):
         Disable when you do not want the resulting plot to be interactive,
         e.g. when trying to display an interactive plot a second time.""")
 
+    precompute = param.Boolean(default=False, doc="""
+       Whether to apply precomputing operations, if available.
```
Comment (jlstevens) [inline holoviews/operation/datashader.py]: Why 'if available'?

Comment (jlstevens) [reply holoviews/operation/datashader.py]: When do you *not* want to precompute? I assume the default is False because precomputation has disadvantages?

Comment (philippjfr) [reply holoviews/operation/datashader.py]: Sure, caching always has disadvantages, namely increased memory usage.

Comment (jlstevens) [reply holoviews/operation/datashader.py]: Ok, then the docstring should mention that.

Comment (philippjfr) [reply holoviews/operation/datashader.py]: >Why 'if available'?

regridding doesn't have or need precomputing.

File: holoviews/operation/datashader.py (hunk 1 above)
Comment (philippjfr) [reply holoviews/operation/datashader.py]: Requires a very specific version of datashader. Could move it to the top in a try/except, up to you.

File: holoviews/operation/datashader.py (hunk 3 above)
Comment (jlstevens) [reply holoviews/operation/datashader.py]: I'm also confused, this is not the sort of caching I was thinking of. I thought you cache from the previous call and if it is the same element, the data is available. If the data changes, then the cache is invalid and cleared.

Maybe you mean the operation *holds* onto memory between calls but from what I can tell, the transient memory usage when the operation is being applied should be the same.

Comment (philippjfr) [reply holoviews/operation/datashader.py]: >I'm also confused, this is not the sort of caching I was thinking of. I thought you cache from the previous call and if it is the same element, the data is available. If the data changes, then the cache is invalid and cleared.

That's correct.

>Maybe you mean the operation holds onto memory between calls but from what I can tell, the transient memory usage when the operation is being applied should be the same.

Correct.

Comment (jlstevens) [reply holoviews/operation/datashader.py]: I think the user doesn't need to know about regridding needing/not needing precomputation - whatever setting there is can be safely ignores so I wouldn't mention it, just to simplify the docstring and minimize confusion.

File: holoviews/operation/datashader.py (hunk 1 above)
Comment (jlstevens) [reply holoviews/operation/datashader.py]: Ok. That is fine I think.

File: holoviews/operation/datashader.py (hunk 3 above)
Comment (philippjfr) [reply holoviews/operation/datashader.py]: Sounds good.

Comment (jlstevens) [reply holoviews/operation/datashader.py]: Maybe something like:

> Whether to apply precomputing operations. Precomputing can speed up resampling operations by avoiding recomputation when the supplied element does not change at the cost of holding onto the memory used for this internal state between calls.

**Comment by jlstevens** [comment]
Other than clarifying the docstring, this PR looks good to me. Happy to merge once it is updated.

**Comment by jlstevens** [comment]
Python 2 builds have passed and we are having some issues with conda so I won't wait for the Python 3 ones. Merging.

**Pull request merged**